File: plugins/dummy_c/dummy.cpp
// SPDX-License-Identifier: Apache-2.0
/*
Copyright (C) 2023 The Falco Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <falcosecurity/sdk.h>

#define PLUGIN_ID 4
#define PLUGIN_NAME "dummy_c"
#define PLUGIN_DESCRIPTION "Reference plugin for educational purposes"
#define PLUGIN_CONTACT "github.com/falcosecurity/plugins"
#define PLUGIN_VERSION "0.2.3"
#define PLUGIN_SOURCE_NAME "dummy_c"

#define PLUGIN_LOG_PREFIX "[dummy_c]"
#define DEFAULT_JITTER 10
#define DEFAULT_MAX_EVENTS 20
#define DEFAULT_START_VALUE 1

class dummy_source
{
    public:
    virtual ~dummy_source() = default;

    dummy_source(uint64_t max_evts, uint64_t start, uint64_t jitter):
            m_event_count(0), m_max_evts(max_evts), m_sample_value(start),
            m_jitter(jitter), m_enc()
    {
    }

    falcosecurity::result_code next_event(falcosecurity::event_writer &evt)
    {
        if(m_event_count >= m_max_evts)
        {
            return falcosecurity::result_code::SS_PLUGIN_EOF;
        }
        m_event_count++;

        // Increment sample by 1, also add a jitter of [0:jitter]
        m_sample_value += 1 + (random() % (m_jitter + 1));

        // we will memcpy the content of `m_sample_value` inside `m_enc.encode`.
        m_enc.set_data((void *)&m_sample_value, sizeof(uint64_t));
        m_enc.encode(evt);
        return falcosecurity::result_code::SS_PLUGIN_SUCCESS;
    }

    private:
    uint64_t m_event_count;
    uint64_t m_max_evts;
    uint64_t m_sample_value;
    uint64_t m_jitter;
    falcosecurity::events::pluginevent_e_encoder m_enc;
};

class dummy
{
    public:
    virtual ~dummy() = default;

    std::string get_name() { return PLUGIN_NAME; }

    std::string get_version() { return PLUGIN_VERSION; }

    std::string get_description() { return PLUGIN_DESCRIPTION; }

    std::string get_contact() { return PLUGIN_CONTACT; }

    uint32_t get_id() { return PLUGIN_ID; };

    std::string get_event_source() { return PLUGIN_SOURCE_NAME; }

    std::string get_last_error() { return m_lasterr; }

    bool init(falcosecurity::init_input &in)
    {
        // Config is optional. In this case, defaults are used.
        if(in.get_config().empty())
        {
            return true;
        }

        try
        {
            auto cfg = nlohmann::json::parse(in.get_config());
            auto it = cfg.find("jitter");
            if(it != cfg.end())
            {
                m_jitter = *it;
            }
        }
        catch(std::exception e)
        {
            m_lasterr = "unable to parse the json config";
            log_error(m_lasterr);
            return false;
        }

        return true;
    }

    std::vector<std::string> get_extract_event_sources()
    {
        return {PLUGIN_SOURCE_NAME};
    }

    std::vector<falcosecurity::field_info> get_fields()
    {
        // We need to compile at least with c++11 to use an ordinary initializer
        // list.
        auto divisibile_arg = falcosecurity::field_arg();
        divisibile_arg.required = true;
        divisibile_arg.index = true;

        using ft = falcosecurity::field_value_type;
        return {
                {ft::FTYPE_UINT64, "dummy.divisible",
                 "Return 1 if the value is divisible by the provided divisor, "
                 "0 otherwise",
                 "Return 1 if the value is divisible by the provided divisor, "
                 "0 otherwise",
                 divisibile_arg},
                {ft::FTYPE_UINT64, "dummy.value",
                 "The sample value in the event",
                 "The sample value in the event"},
                {ft::FTYPE_STRING, "dummy.strvalue",
                 "The sample value in the event, as a string",
                 "The sample value in the event, as a string"},
        };
    }

    void log_error(std::string err_mess)
    {
        printf("%s %s\n", PLUGIN_LOG_PREFIX, err_mess.c_str());
    }

    bool extract(const falcosecurity::extract_fields_input &in)
    {
        auto &req = in.get_extract_request();
        falcosecurity::events::pluginevent_e_decoder dec(in.get_event_reader());
        uint32_t len = 0;
        auto sample = *((uint64_t *)(dec.get_data(len)));
        if(len != sizeof(uint64_t))
        {
            log_error("invalid event payload");
        }

        switch(req.get_field_id())
        {
        case 0: // dummy.divisible
        {
            if(!req.is_arg_present())
            {
                log_error("'dummy.divisible' requires an argument but no "
                          "argument is provided");
                return false;
            }

            uint64_t res = 0;
            auto divisor = req.get_arg_index();
            if((sample % divisor) == 0)
            {
                res = 1;
            }
            req.set_value(res, true);
            return true;
        }
        case 1: // dummy.value
        {
            req.set_value(sample, true);
            return true;
        }
        case 2: // dummy.strvalue
        {
            // The event payload is simply the sample, as a string
            req.set_value(std::to_string(sample), true);
            return true;
        }
        default:
            m_lasterr = "no known field: " + std::to_string(req.get_field_id());
            log_error(m_lasterr);
            return false;
        }

        return false;
    }

    std::unique_ptr<dummy_source> open(const std::string &params)
    {
        // Config is optional. In this case, defaults are used.
        uint64_t max_events = DEFAULT_MAX_EVENTS;
        uint64_t start = DEFAULT_START_VALUE;

        if(!params.empty())
        {
            try
            {
                auto open_params = nlohmann::json::parse(params);
                auto it = open_params.find("start");
                if(it != open_params.end())
                {
                    start = *it;
                }

                it = open_params.find("maxEvents");
                if(it != open_params.end())
                {
                    max_events = *it;
                }
            }
            catch(std::exception e)
            {
                m_lasterr = "wrong open params format";
                log_error(m_lasterr);
                return nullptr;
            }
        }

        return std::unique_ptr<dummy_source>(
                new dummy_source(max_events, start, m_jitter));
    }

    std::string m_lasterr = "";
    uint64_t m_jitter = DEFAULT_JITTER;
};

FALCOSECURITY_PLUGIN(dummy);
FALCOSECURITY_PLUGIN_EVENT_SOURCING(dummy, dummy_source);
FALCOSECURITY_PLUGIN_FIELD_EXTRACTION(dummy);

File: plugins/dummy/pkg/dummy/dummy.go
// SPDX-License-Identifier: Apache-2.0
/*
Copyright (C) 2024 The Falco Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package dummy

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"strconv"
	"time"

	"github.com/alecthomas/jsonschema"
	"github.com/falcosecurity/plugin-sdk-go/pkg/sdk"
	"github.com/falcosecurity/plugin-sdk-go/pkg/sdk/plugins"
	"github.com/falcosecurity/plugin-sdk-go/pkg/sdk/plugins/source"
)

const (
	PluginID          uint32 = 3
	PluginName               = "dummy"
	PluginDescription        = "Reference plugin for educational purposes"
	PluginContact            = "github.com/falcosecurity/plugins"
	PluginVersion            = "0.11.4"
	PluginEventSource        = "dummy"
)

type PluginConfig struct {
	// This reflects potential internal state for the plugin. In
	// this case, the plugin is configured with a jitter.
	Jitter uint64 `json:"jitter" jsonschema:"title=Sample jitter,description=A random amount added to the sample of each event (Default: 10),default=10"`
}

type PluginOpenParams struct {
	Start     uint64 `json:"start" jsonschema:"title=Start value,description=The starting value of the sample (Default: 1),default=1"`
	MaxEvents uint64 `json:"maxEvents" jsonschema:"title=Max num events,description=The number of events to return before returning EOF (Default: 20),default=20"`
}

type Plugin struct {
	plugins.BasePlugin
	// Will be used to randomize samples
	rand *rand.Rand
	// Contains the init configuration values
	config PluginConfig
	// Contains the open params configuration
	openParams PluginOpenParams
}

func (p *PluginConfig) setDefault() {
	p.Jitter = 10
}

func (p *PluginOpenParams) setDefault() {
	p.Start = 1
	p.MaxEvents = 20
}

func (m *Plugin) Info() *plugins.Info {
	return &plugins.Info{
		ID:          PluginID,
		Name:        PluginName,
		Description: PluginDescription,
		Contact:     PluginContact,
		Version:     PluginVersion,
		EventSource: PluginEventSource,
	}
}

func (p *Plugin) InitSchema() *sdk.SchemaInfo {
	reflector := jsonschema.Reflector{
		RequiredFromJSONSchemaTags: true, // all properties are optional by default
		AllowAdditionalProperties:  true, // unrecognized properties don't cause a parsing failures
	}
	if schema, err := reflector.Reflect(&PluginConfig{}).MarshalJSON(); err == nil {
		return &sdk.SchemaInfo{
			Schema: string(schema),
		}
	}
	return nil
}

func (p *Plugin) Init(cfg string) error {
	// initialize state
	p.rand = rand.New(rand.NewSource(time.Now().UnixNano()))

	// The format of cfg is a json object with a single param
	// "jitter", e.g. {"jitter": 10}
	// Empty configs are allowed, in which case the default is used.
	// Since we provide a schema through InitSchema(), the framework
	// guarantees that the config is always well-formed json.
	p.config.setDefault()
	if len(cfg) != 0 {
		json.Unmarshal([]byte(cfg), &p.config)
	}
	return nil
}

func (p *Plugin) Destroy() {
	// nothing to do here
}

func (p *Plugin) Open(prms string) (source.Instance, error) {

	p.openParams.setDefault()
	if len(prms) != 0 {
		if err := json.Unmarshal([]byte(prms), &p.openParams); err != nil {
			return nil, fmt.Errorf("wrong open params format: %s", err.Error())
		}
	}

	evt_counter := uint64(0)
	sample := p.openParams.Start
	maxEvents := p.openParams.MaxEvents
	pull := func(ctx context.Context, evt sdk.EventWriter) error {
		if evt_counter >= uint64(maxEvents) {
			return sdk.ErrEOF
		}
		evt_counter++

		// Increment sample by 1, also add a jitter of [0:jitter]
		sample += 1 + uint64(p.rand.Int63n(int64(p.config.Jitter+1)))

		// The representation of a dummy event is the sample as a string.
		str := fmt.Sprintf("%d", sample)

		// It is not mandatory to set the Timestamp of the event (it
		// would be filled in by the framework if set to uint_max),
		// but it's a good practice.
		evt.SetTimestamp(uint64(time.Now().UnixNano()))

		_, err := evt.Writer().Write([]byte(str))
		return err
	}
	return source.NewPullInstance(pull)
}

// todo: optimize this to cache by event number
func (m *Plugin) String(evt sdk.EventReader) (string, error) {
	evtBytes, err := ioutil.ReadAll(evt.Reader())
	if err != nil {
		return "", err
	}
	evtStr := string(evtBytes)

	// The string representation of an event is a json object with the sample
	return fmt.Sprintf("{\"sample\": \"%s\"}", evtStr), nil
}

func (m *Plugin) Fields() []sdk.FieldEntry {
	return []sdk.FieldEntry{
		{
			Type: "uint64",
			Name: "dummy.divisible",
			Desc: "Return 1 if the value is divisible by the provided divisor, 0 otherwise",
			Arg:  sdk.FieldEntryArg{IsRequired: true, IsIndex: true},
		},
		{
			Type: "uint64",
			Name: "dummy.value",
			Desc: "The sample value in the event",
		},
		{
			Type: "string",
			Name: "dummy.strvalue",
			Desc: "The sample value in the event, as a string",
		},
	}
}

func (m *Plugin) Extract(req sdk.ExtractRequest, evt sdk.EventReader) error {
	evtBytes, err := ioutil.ReadAll(evt.Reader())
	if err != nil {
		return err
	}
	evtStr := string(evtBytes)
	evtVal, err := strconv.Atoi(evtStr)
	if err != nil {
		return err
	}

	switch req.FieldID() {
	case 0: // dummy.divisible

		if !req.ArgPresent() {
			return fmt.Errorf("'dummy.divisible' field requires an argument, but no argument is provided")
		}

		divisor := req.ArgIndex()

		if uint64(evtVal)%divisor == 0 {
			req.SetValue(uint64(1))
		} else {
			req.SetValue(uint64(0))
		}
	case 1: // dummy.value
		req.SetValue(uint64(evtVal))
	case 2: // dummy.strvalue
		req.SetValue(evtStr)
	default:
		return fmt.Errorf("no known field: %s", req.Field())
	}

	return nil
}
